src/containers/rbtree/rbtree.h

```diff
@@ -0,0 +1,794 @@
+      source->parent_->left_ = target;
+    } else {
+      source->parent_->right_ = target;
+    }
+  }
+
+  if (target->parent_) {
+    if (target->parent_->left_ == target) {
+      target->parent_->left_ = source;
+    } else {
+      target->parent_->right_ = source;
+    }
+  }
+
+  tempNode = source->parent_;
+  source->parent_ = target->parent_;
+  target->parent_ = tempNode;
+}
+
+template <typename K, typename T>
+inline void rbtree<K, T>::replace_node(node* old_node,
+                                       node* new_node) noexcept {
+  if (new_node) {
+    new_node->parent_ = old_node->parent_;
+  }
+
+  if (old_node == old_node->parent_->left_) {
+    old_node->parent_->left_ = new_node;
+  } else {
+    old_node->parent_->right_ = new_node;
+  }
+}
+
+template <typename K, typename T>
+inline void rbtree<K, T>::delete_case1(node* n) {
+  if (n->parent_) {
+    delete_case2(n);
+  }
+}
+
+template <typename K, typename T>
+inline void rbtree<K, T>::delete_case2(node* find) {
+  node* s = sibling(find);
+  if (s && s->mColor == 'R') {
+    find->parent_->mColor = 'R';
+    s->mColor = 'B';
+    if (find == find->parent_->left_) {
+      rotate_left(find->parent_);
+    } else {
+      rotate_right(find->parent_);
+    }
+  }
+  delete_case3(find);
+}
+
+template <typename K, typename T>
+inline void rbtree<K, T>::delete_case3(node* find) {
+  node* s = sibling(find);
+  if (s && find->parent_->mColor == 'B' && s->mColor == 'B' &&
+      (!s->left_ || s->left_->mColor == 'B') &&
+      (!s->right_ || s->right_->mColor == 'B')) {
+    s->mColor = 'R';
+    delete_case1(find->parent_);
+  } else {
+    delete_case4(find);
+  }
+}
+
+template <typename K, typename T>
+inline void rbtree<K, T>::delete_case4(node* n) {
+  node* s = sibling(n);
+  if (s && n->parent_->mColor == 'R' && (s->mColor == 'B') &&
+      (!s->left_ || s->left_->mColor == 'B') &&
+      (!s->right_ || s->right_->mColor == 'B')) {
+    s->mColor = 'R';
+    n->parent_->mColor = 'B';
+  } else {
+    delete_case5(n);
+  }
+}
+
+template <typename K, typename T>
+inline void rbtree<K, T>::delete_case5(node* n) {
+  node* s = sibling(n);
+
+  if (s && s->mColor == 'B') {
+    if ((n == n->parent_->left_) && (!s->right_ || s->right_->mColor == 'B') &&
+        (!s->left_ || s->left_->mColor == 'R')) {
+      s->mColor = 'R';
+      s->left_->mColor = 'B';
+      rotate_right(s);
+    } else if ((n == n->parent_->right_) && (s->left_->mColor == 'B') &&
+               (s->right_->mColor == 'R')) {
+      s->mColor = 'R';
+      s->right_->mColor = 'B';
+      rotate_left(s);
+    }
+  }
+  delete_case6(n);
+}
+
+template <typename K, typename T>
+inline void rbtree<K, T>::delete_case6(node* n) {
+  node* s = sibling(n);
+  s->mColor = n->parent_->mColor;
+  n->parent_->mColor = 'B';
+  if (n == n->parent_->left_) {
+    s->right_->mColor = 'B';
+    rotate_left(n->parent_);
+  } else {
+    s->left_->mColor = 'B';
+    rotate_right(n->parent_);
+  }
+}
```

Comment:
1. `sibling()` может вернуть `nullptr`, но на `nullptr` мы не проверяем и пытаемся сразу обратиться к `s`. В случае если `sibling()` вернёт `nullptr` эта функция крашнет программу с ошибкой `SIGSEGV`.
2. Лучше полностью называть сущности. Вместо `s` лучше `sibling`. Собственно, при таком подходе сразу захочется явно выделять методы как методы и добавлять в их название глаголы, например, как писал выше - `get_sibling()`, вместо `sibling()`.

```suggestion
inline void rbtree<K, T>::delete_case6(node* n) {
  node* sibling = get_sibling(n);
  if (!sibling) return;

  sibling->mColor = n->parent_->mColor;
  n->parent_->mColor = 'B'; // Не забудь переделать на enum

  if (n == n->parent_->left_) {
    if (sibling->right_) sibling->right_->mColor = 'B';
    rotate_left(n->parent_);
  } else {
    if (sibling->left_) sibling->left_->mColor = 'B';
    rotate_right(n->parent_);
  }
}
```